File: internal/sites/bilibili/grpc.go
package bilibili

import (
	"bytes"
	"compress/gzip"
	"encoding/binary"
	"io"

	"github.com/ac1982/haul/internal/errs"
)

// gRPC length-prefixed framing: a compressed flag byte, a 4-byte big-endian length, then the message. The APP API
// is plain HTTPS POSTs carrying one such frame each way.

// packFrame gzips message into a compressed frame.
func packFrame(message []byte) []byte {
	var body bytes.Buffer
	zw := gzip.NewWriter(&body)
	_, _ = zw.Write(message) // writes to a bytes.Buffer cannot fail
	_ = zw.Close()
	out := make([]byte, 5, 5+body.Len())
	out[0] = 1
	binary.BigEndian.PutUint32(out[1:], uint32(body.Len()))
	return append(out, body.Bytes()...)
}

// unpackFrame is the message of the first frame; bytes past its declared length (trailers) are ignored.
func unpackFrame(data []byte) ([]byte, error) {
	if len(data) < 5 {
		return nil, errs.New("Incomplete gRPC response (%d bytes)", len(data))
	}
	body := data[5:]
	if n := binary.BigEndian.Uint32(data[1:5]); uint64(n) < uint64(len(body)) {
		body = body[:n]
	}
	if data[0] != 1 {
		return body, nil
	}
	zr, err := gzip.NewReader(bytes.NewReader(body))
	if err != nil {
		return nil, errs.New("Invalid gRPC response: %v", err)
	}
	out, err := io.ReadAll(zr)
	if err != nil {
		return nil, errs.New("Invalid gRPC response: %v", err)
	}
	return out, nil
}

File: Sources/HaulCore/Sites/Bilibili/GRPC.swift
import Foundation

/// gRPC length-prefixed message framing: 1 byte compressed flag + 4 byte big-endian length + payload.
public enum GRPCFrame {
    public static func pack(_ message: Data) throws -> Data {
        let payload = try Gzip.compress(message)
        var out = Data(capacity: 5 + payload.count)
        out.append(1)
        var len = UInt32(payload.count).bigEndian
        withUnsafeBytes(of: &len) { out.append(contentsOf: $0) }
        out.append(payload)
        return out
    }

    public static func unpack(_ data: Data) throws -> Data {
        guard data.count >= 5 else { throw HaulError("Incomplete gRPC response (\(data.count) bytes)") }
        let start = data.startIndex
        let flag = data[start]
        let len = data[(start + 1)..<(start + 5)].withUnsafeBytes { UInt32(bigEndian: $0.loadUnaligned(as: UInt32.self)) }
        let end = min(data.endIndex, start + 5 + Int(len))
        let body = data.subdata(in: (start + 5)..<end)
        return flag == 1 ? try Gzip.decompress(body) : body
    }
}
